File: daily-exercise/special_topic/link/reverse-linked-list-ii.go
//Definition for singly-linked list.
type ListNode struct {
    Val int
    Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
    // 虚拟头节点
    preHead := &ListNode{Next:head}
    // 左前头节点
    leftBefore := preHead
    for i := 0;i<left-1;i++{
        leftBefore = leftBefore.Next
    }
    // 左右节点
    leftNode,rightNode := leftBefore.Next,leftBefore.Next
    for i:=0;i<right-left;i++{
        rightNode = rightNode.Next
    }
    // 右后节点
    rightAfter := rightNode.Next
    // 切断连接
    leftBefore.Next = nil
    rightNode.Next = nil
    // 翻转局部链表
    reverseLink(leftNode)
    // 建立新连接
    leftBefore.Next = rightNode
    leftNode.Next = rightAfter
    return preHead.Next
}
func reverseLink(head *ListNode) {
    var pre *ListNode
    cur := head
    for cur != nil {
        nex := cur.Next
        cur.Next = pre
        pre, cur = cur, nex
    }
}

File: daily-exercise/special_topic/link/reverse-linked-list-ii.py
# -*- coding: utf-8 -*-
# @Time    : 2022/8/14 15:13
# @Author  : kevin
# @Version : python 3.7
# @Desc    : reverse-linked-list-ii

# 翻转链表2
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 方法：拿到left-right区间链表断开连接，局部翻转（调用翻转方法），拿到左前，右后节点，重新建立新连接即可
class Solution:
    def reverseBetween(self, head, left, right):
        if not head:
            return None
        # 虚拟头
        preHead = ListNode(next=head)
        leftBefore = preHead
        # 左前节点
        for _ in range(left - 1):
            leftBefore = leftBefore.next
        # 创建左右区间链表
        leftNode, rightNode = leftBefore.next, leftBefore
        for _ in range(right - left + 1):
            rightNode = rightNode.next
        # 断开连接
        rightAfter = rightNode.next
        leftBefore.next = None
        rightNode.next = None
        # 翻转链表
        self.reverse_link(leftNode)
        # 简历新连接
        leftBefore.next = rightNode
        leftNode.next = rightAfter

        return preHead.next

    def reverse_link(self, head):
        pre, cur = None, head
        while cur:
            nex = cur.next
            cur.next = pre
            pre, cur = cur, nex
